ext/standard/basic_functions.stub.php

```diff
@@ -1510,3 +1512,18 @@ function sapi_windows_set_ctrl_handler(?callable $handler, bool $add = true): bo
 
 function sapi_windows_generate_ctrl_event(int $event, int $pid = 0): bool {}
 #endif
+} // end global namespace
+
+namespace iterable {
+
+function any(iterable $iterable, ?callable $callback = null): bool {}
+
+function all(iterable $iterable, ?callable $callback = null): bool {}
+
+function none(iterable $iterable, ?callable $callback = null): bool {}
+
+function reduce(iterable $iterable, callable $callback, mixed $initial = null): mixed {}
+
+function find(iterable $iterable, callable $callback, mixed $default = null): mixed {}
```

Comment:
Can we remove this one? I'd rather add filter, map, and flatmap if we need more functions.

Reply:
find() is also useful and many other programming languages include both (js, haskell, etc.)

For example, if you have an array of a million elements and only want the first match, it is much more efficient to call find() if the iterable contains a matching value (and there would be less service calls and db calls) compared to `reset(array_filter(...))`

Additionally, filter() and map() would be waiting on the existence of CachedIterable when it starts, because Traversables can have repeated keys `yield 'key' => 1; yield 'key' => 2;`

Reply:
> Additionally, filter() and map() would be waiting on the existence of CachedIterable when it starts, because Traversables can have repeated keys `yield 'key' => 1; yield 'key' => 2;`

This does not require a CachedIterable; they can return any iterator.

Reply:
I mean that there's nothing in the standard library yet that supports rewinding, counting, and especially arbitrary offset access

It seems much more difficult to use without the support for rewindability, random offset access, countability, etc.

But yes, I suppose you could hide the implementation entirely with InternalIterator and only support a single iteration

Reply:
> Edit: on second thought, they should not return a cached iterable. These routines are often chained together; if every piece of the chain caches their results, it will balloon memory usage.

CachedIterable is basically the name chosen for a rewindable immutable key-value sequence. It isn't cached permanently, it has a regular object lifetime. I'm referring to https://wiki.php.net/rfc/cachediterable

Reply:
On second thought, `filter` and `map` should not return a CachedIterable. These routines are often chained together with other iterable routines; if every piece of the chain caches their results, it will balloon memory usage.

Reply:
> CachedIterable is basically the name chosen for a rewindable immutable key-value sequence. It isn't cached permanently, it has a regular object lifetime. I'm referring to https://wiki.php.net/rfc/cachediterable

Yes, I am specifically saying they _should not_ return this object. It will ballon memory usage to do it that way. Think about it; you have a filter + map plus some terminator like `first_n` with n=100. Filter and map will hold at least 100 values in memory that shouldn't be there while calculating the `first_n`.

Reply:
> Yes, I am specifically saying they should not return this object. It will ballon memory usage to do it that way. Think about it; you have a map + filter plus some terminator like `first_n` with n=100. Filter and map will each hold at least 100 values in memory that shouldn't be there while calculating the `first_n`.

1) Many end users would want the standard library to return something that could be iterated over multiple times - this was a source of bugs in spl classes such as https://www.php.net/arrayobject prior to them being switched to iterables

```
$foo = map(...);
foreach ($foo as $i => $v1) {
    foreach ($foo as $i => $v2) {
        if (some_pair_predicate($v1, $v2)) {
            // do something
        }
    }
}

2) Library/application authors that are interested in lazy generators could use or implement something such as https://github.com/nikic/iter instead - my opinion is that the standard library should provide something that is easy to understand, debug, serialize or represent, etc.
3) It would be harder to understand why SomeFrameworkException is thrown in code unrelated to that framework when it is passed to some function that accepts a generic iterable, and harder to write correct exception handling for it if done in a lazy generation style
4) It is possible to implement a lazy version of CachedIterable that only loads values as needed - but I hadn't proposed it due to uncertainty that 2/3 of voters would consider it widely useful enough to be included in php rather than as a userland or PECL library

Reply:
Iterables are made from arrays and generators. Since there is already a large body of functions which work with arrays, it seems reasonable to assume they would use an `Iterable` part of the standard library because they also use iterators/generators. One of the main points of generators is to _reduce_ memory. We shouldn't add a corpus of iterable functions that will increase memory; that works directly against the goals of the iterable feature!

No, it's much better to compose something if you want it to be eager, e.g. `$foo = CachedIterable::new(map(...$args));`

Reply:
> Iterables are made from arrays and generators

Iterables are made from arrays **and Traversable objects such as** SplObjectStorage, ArrayObject, user-defined data structures, and Generators

> We shouldn't add a corpus of iterable functions that will increase memory; that works directly against the goals of the iterable feature!

In common use cases, the memory increase may be small, especially if the lifetime of the variable or temporary expression holding the result is short.

Additionally, a lazy iterable would require keeping a reference to the previous (possibly lazy) iterable, and the iterables/arrays those reference - if the initial iterable is larger than the result then eagerly evaluating `$var = map($cb, temporary())` would end up saving memory

> No, it's much better to compose something if you want it to be eager, e.g. `$foo = CachedIterable::new(map(...$args));` We can include this in the examples in the docs for filter, map, etc.

That seems error prone and I'm personally opposed to that.

PHP has typically been imperitive rather than functional, and focused on "cater[ing] to the skill-levels and platforms of a wide range of users" as RFC authors are repeatedly reminded in https://wiki.php.net/rfc/template and my interpretation of that is that imperitive would be much more acceptable (aside: The loosely typed language part seems less applicable nowadays)

> PHP is and should remain:
> 1) a pragmatic web-focused language
> 2) a loosely typed language
> 3) a language which caters to the skill-levels and platforms of a wide range of users

Lazy data structures would be easy to misuse (consume twice, attempt to serialize or encode, attempt to log the full iterable(consume twice) etc) without (or even with) linters and static analyzers, so this really doesn't seem like catering to a wide range of users.

-------

Explicitly using a different family of functions to act on generators internally would probably make more sense than being the default, e.g. https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#findFirst-- and https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html (Streams are separate from java.util.Collection in java, javascript eagerly evaluates https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Map, etc)

Reply:
> No, it's much better to compose something if you want it to be eager, e.g. `$foo = CachedIterable::new(map(...$args));` We can include this in the examples in the docs for filter, map, etc.

Having the eager version be longer than the lazy version (instead of shorter or the same length) would also encourage the use of the lazy version, which I'd objected to for being error prone and easy to misuse.